### PR TITLE
 improved   speed of mean_backward under certain condition

### PR DESCRIPTION
**Pull request by abh2k**
https://colab.research.google.com/drive/1dqjdHB4-Rw1HNDd59gFFA468WEPY7oTf?usp=sharing

### CODE REVIEW

File: src/mlpack/methods/ann/layer/mean_pooling.hpp [hunk 1]
```diff
@@ -195,35 +195,86 @@ class MeanPooling
...
   {
 
-    arma::Mat<eT> unpooledError;
-    for (size_t j = 0, colidx = 0; j < input.n_cols; j += strideHeight, colidx++)
-    {
-      for (size_t i = 0, rowidx = 0; i < input.n_rows; i += strideWidth, rowidx++)
-      { 
-        size_t rowEnd = i + kernelWidth - 1;
-        size_t colEnd = j + kernelHeight - 1;
+    size_t condition = kernelHeight * kernelWidth - strideHeight * strideWidth -
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/mean_pooling.hpp:198]: We could make this const.

File: src/mlpack/methods/ann/layer/mean_pooling.hpp [hunk 2]
```diff
@@ -195,35 +195,86 @@ class MeanPooling
...
+      kernelWidth - kernelHeight;
 
-        if (rowEnd > input.n_rows - 1)
+    if (condition > 0)
+    {
+      for (size_t j = 0, colidx = 0; j < input.n_cols; j += strideHeight, colidx++)
+      {
+        for (size_t i = 0, rowidx = 0; i < input.n_rows; i += strideWidth, rowidx++)
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/mean_pooling.hpp:205]:
```suggestion
      for (size_t j = 0, colidx = 0; j < input.n_cols; j += strideHeight, ++colidx)
      {
        for (size_t i = 0, rowidx = 0; i < input.n_rows; i += strideWidth, ++rowidx)
```

**Review by zoq — COMMENTED** [review]
Do you think you could add some comments, I find this somewhat difficult to follow.

**Comment by abh2k** [comment]
> Do you think you could add some comments, I find this somewhat difficult to follow.

Done. I will share a detailed calculation of how I came to the condition .

File: src/mlpack/methods/ann/layer/mean_pooling.hpp [hunk 3]
```diff
@@ -195,35 +195,131 @@ class MeanPooling
...
+      // In prefix method the following operations will be performed:
+      //    1. Add `+e` to input(x1, y1).
+      //    2. Add `-e` to input(x1 + 1, y1).
+      //    3. Add `-e` to input(x1, y1 + 1).
+      //    4. Add `+e` to input(x1 + 1, y1 + 1).
+      //    5. Perform Prefix sum over columns i.e input(i, j) += input(i, j - 1)
+      //    6. Perform Prefix sum over rows i.e input(i, j) += input(i - 1, j)
+      // So lets say if we had `k` number of such operations. The brute force
+      // method will require `kernalArea * k` operations.
+      // The prefix method will require `4 * k + Prefix operation`.
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/mean_pooling.hpp:234]: Great, thanks for adding that.

**Comment by zoq** [comment]
> > Do you think you could add some comments, I find this somewhat difficult to follow.
> 
> Done. I will share a detailed calculation of how I came to the condition .

NIce, with that do you mean the comments you added in the last commit, or something else?

**Comment by zoq** [comment]
I see, yes would be great if you could add a comment for that line as well.

**Comment by abh2k** [comment]
@zoq  I added a better condition which is more  understandable  and its explanation is also added.

File: src/mlpack/methods/ann/layer/mean_pooling.hpp [hunk 4]
```diff
@@ -210,38 +210,139 @@ class MeanPooling
                  const arma::Mat<eT>& error,
                  arma::Mat<eT>& output)
   {
-    arma::Mat<eT> unpooledError;
-    for (size_t j = 0, colidx = 0; j < input.n_cols; j += strideHeight,
-         colidx++)
+    // This condition comes by comparing the number of operations involved in the brute
+    // force method and the prefix method. Let the area of error be errorArea and area
+    // of kernal be kernalArea. Total number of operations in brute force method will be
+    // `errorArea * kernalArea` and for each element in error we are doing kernalArea
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/mean_pooling.hpp:216]:
```suggestion
    // `errorArea * kernalArea` and for each element in error we are doing `kernalArea`
```

File: src/mlpack/methods/ann/layer/mean_pooling.hpp [hunk 5]
```diff
@@ -210,38 +210,139 @@ class MeanPooling
...
+        (4 * error.n_elem + 2 * input.n_elem);
+
+    if (condition)
     {
-      for (size_t i = 0, rowidx = 0; i < input.n_rows; i += strideWidth,
-           rowidx++)
+      // If this condition is true then theoritically the prefix sum method of
+      // unpooling is faster. The aim of unpooling is to add
+      // `error(i, j) / kernalArea` to `inputArea(kernal)`. This requires
+      // inputArea.n_elem additions. So, total operations required will be
```
Comment (zoq) [inline src/mlpack/methods/ann/layer/mean_pooling.hpp:229]:
```suggestion
      // `inputArea.n_elem` additions. So, total operations required will be
```

**Review by zoq — COMMENTED** [review]
Thanks for adding the comment, makes it very clear.

**Review by zoq — APPROVED** [review]
Great, no more comments from my side.

**Comment by zoq** [comment]
Thanks for working on this.

**Pull request merged**